framework/src/meshgenerators/FancyExtruderGenerator.C

```diff
@@ -0,0 +1,547 @@
+                             "each elevation, variable number of layers within each elevation and "
+                             "remap subdomain_ids within each elevation");
+
+  params.addRequiredParam<std::vector<Real>>("heights", "The height of each elevation");
+
+  params.addRequiredParam<std::vector<unsigned int>>(
+      "num_layers", "The number of layers for each elevation - must be num_elevations in length!");
+
+  params.addParam<std::vector<std::vector<subdomain_id_type>>>(
+      "subdomain_swaps",
+      "For each row, every two entries are interpreted as a pair of "
+      "'from' and 'to' to remap the subdomains for that elevation");
+
+  params.addRequiredParam<Point>(
+      "direction",
+      "A vector that points in the direction to extrude (note, this will be "
+      "normalized internally - so don't worry about it here)");
+
+  params.addParam<boundary_id_type>(
+      "top_boundary",
+      "The boundary ID to set on the top boundary.  If ommitted one will be generated.");
+
+  params.addParam<boundary_id_type>(
+      "bottom_boundary",
+      "The boundary ID to set on the bottom boundary.  If omitted one will be generated.");
+
+  return params;
+}
+
+FancyExtruderGenerator::FancyExtruderGenerator(const InputParameters & parameters)
+  : MeshGenerator(parameters),
+    _input(getMesh("input")),
+    _heights(getParam<std::vector<Real>>("heights")),
+    _num_layers(getParam<std::vector<unsigned int>>("num_layers")),
+    _subdomain_swaps(getParam<std::vector<std::vector<subdomain_id_type>>>("subdomain_swaps")),
+    _direction(getParam<Point>("direction")),
+    _has_top_boundary(isParamValid("top_boundary")),
+    _top_boundary(isParamValid("top_boundary") ? getParam<boundary_id_type>("top_boundary") : 0),
+    _has_bottom_boundary(isParamValid("bottom_boundary")),
+    _bottom_boundary(isParamValid("bottom_boundary") ? getParam<boundary_id_type>("bottom_boundary")
+                                                     : 0)
+{
+  // Normalize it
+  _direction /= _direction.norm();
+
+  const auto num_elevations = _heights.size();
+
+  if (_num_layers.size() != num_elevations)
+    mooseError("The length of 'heights' and 'num_layers' must be the same in ", name());
+
+  if (_subdomain_swaps.size() && (_subdomain_swaps.size() != num_elevations))
+    mooseError("If specified, 'subdomain_swaps' must be the same length as 'heights' in ", name());
+
+  _subdomain_swap_pairs.resize(_subdomain_swaps.size());
+
+  // Reprocess the subdomain swaps to make pairs out of them so they are easier to use
+  for (MooseIndex(_subdomain_swaps) i = 0; i < _subdomain_swaps.size(); i++)
+  {
+    const auto & elevation_swaps = _subdomain_swaps[i];
+    auto & elevation_swap_pairs = _subdomain_swap_pairs[i];
+
+    for (MooseIndex(elevation_swaps) i = 0; i < elevation_swaps.size(); i += 2)
+      elevation_swap_pairs[elevation_swaps[i]] = elevation_swaps[i + 1];
```

Comment:
Need to do sanity checking on your `subdomain_swaps` input, if a user puts in an odd number you'll have an out of bounds access.

Reply:
Also need sanity checking on your swap pairs. If somebody puts in too large of a number, you'll have a separate out of bounds access.

Reply:
?  I don't see that one.  The size is set by the number of swaps...

BUT: the swaps do need to be _even_